gateway/src/http_entrypoints.rs

```diff
@@ -0,0 +1,446 @@
+            flt_a2: state.flags.intersects(IgnitionFlags::FLT_A2),
+            flt_rot: state.flags.intersects(IgnitionFlags::FLT_ROT),
+            flt_sp: state.flags.intersects(IgnitionFlags::FLT_SP),
+        }
+    }
+}
+
+#[derive(Serialize, JsonSchema)]
+struct SpComponentInfo;
+
+#[derive(Deserialize, JsonSchema)]
+#[allow(dead_code)] // TODO remove once this is used
+struct Timeout {
+    timeout: Option<u32>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct TimeoutSelector<T> {
+    last: T,
+    start_time: u64, // TODO
+}
+
+#[derive(Serialize, Deserialize, JsonSchema, PartialEq, Debug, Clone, Copy)]
+#[serde(rename_all = "lowercase")]
+pub(crate) enum SpType {
+    Sled,
+    Power,
+    Switch,
+}
+
+#[derive(Serialize, Deserialize, JsonSchema, PartialEq, Debug, Clone)]
+pub(crate) struct SpIdentifier {
+    #[serde(rename = "type")]
+    pub(crate) typ: SpType,
+    pub(crate) slot: u32,
+}
+
+impl SpIdentifier {
+    fn placeholder_map_to_target(
+        &self,
+        known_sps: &KnownSps,
+    ) -> Result<u8, Error> {
+        // TODO This is wrong in all kinds of ways, but is just a placeholder
+        // for now until we have a better story for bootstrapping how MGS knows
+        // which SP is which.
+        //
+        // Maps `self` to a target number by assuming target numbers are indexed
+        // from 0 starting with switches followed by sleds followed by power
+        // controllers.
+
+        let slot: usize = usize::try_from(self.slot)
+            .map_err(|_| Error::SpDoesNotExist(self.clone()))?;
+
+        let mut base = 0;
+        for (typ, count) in [
+            (SpType::Switch, known_sps.switches.len()),
+            (SpType::Sled, known_sps.sleds.len()),
+            (SpType::Power, known_sps.power_controllers.len()),
+        ] {
+            if self.typ != typ {
+                base += count;
+                continue;
+            }
+
+            if slot < count {
+                let slot = u8::try_from(slot + base).map_err(|_| {
+                    Error::InternalError {
+                        internal_message:
+                            "too many total configured SP slots (must be < 256)"
+                                .to_string(),
+                    }
+                })?;
+                return Ok(slot);
+            } else {
+                return Err(Error::SpDoesNotExist(self.clone()));
+            }
+        }
+
+        // above loop returns once we match on `typ`
+        unreachable!()
+    }
+}
+
+type TimeoutPaginationParams<T> = PaginationParams<Timeout, TimeoutSelector<T>>;
+
+#[derive(Deserialize, JsonSchema)]
+struct PathSp {
+    /// ID for the SP that the gateway service translates into the appropriate
+    /// port for communicating with the given SP.
+    ///
+    /// This should be a `#[serde(flatten)]`'d [`SpIdentifier`], but serde
+    /// cannot deserialize flattened types containing non-`String` types through
+    /// a dropshot [`Path`].
```

Comment:
give this, I suggest we just do away with the `SpIdentifier` type

Reply:
I think this is entirely stylistic, but I'm not sure I agree - `SpIdentifier` is used in a couple places where `PathSp` doesn't seem to be as nice of a name (e.g., I used it in `Error::SpDoesNotExist(_)`, and you used it inside a `TimeoutPaginationParams<_>`). Could we get rid of `PathSp` instead? Is `Path<SpIdentifier>` unidiomatic dropshot?

Reply:
I found this annoying enough I took a stab at manually implementing `Deserialize` in a way that allows `flatten` to work the way you thought it should originally. Code isn't pretty, but I think it's pretty normal by serde standards: cd3cdde